File: mcp-code-intelligence-python/src/mcp_code_intel/graph/symbol_resolver.py
"""Symbol Resolver — resolves symbol names to database records. KSA-179."""

from __future__ import annotations

import sqlite3

from .models import ResolvedSymbol


class SymbolResolver:
    """Resolves symbol identifiers to indexed symbol records.

    Supports: exact match, qualified names (Class.method), file:symbol format.
    """

    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def resolve(self, input_name: str) -> list[ResolvedSymbol]:
        """Resolve a symbol name to one or more database records."""
        # Strategy 1: Exact match
        results = self._exact_match(input_name)
        if results:
            return results

        # Strategy 2: Qualified name (Class.method)
        if "." in input_name:
            dot_idx = input_name.rfind(".")
            parent = input_name[:dot_idx]
            method = input_name[dot_idx + 1:]
            results = self._qualified_match(method, parent)
            if results:
                return results

        # Strategy 3: file:symbol format
        if ":" in input_name:
            colon_idx = input_name.rfind(":")
            file_part = input_name[:colon_idx]
            name_part = input_name[colon_idx + 1:]
            results = self._file_match(name_part, file_part)
            if results:
                return results

        return []

    def suggest(self, input_name: str, limit: int = 5) -> list[str]:
        """Suggest similar symbol names for 'did you mean?' responses."""
        cur = self._conn.execute(
            "SELECT DISTINCT name FROM symbols WHERE name LIKE ? LIMIT ?",
            (f"%{input_name}%", limit),
        )
        return [row[0] for row in cur.fetchall()]

    def _exact_match(self, name: str) -> list[ResolvedSymbol]:
        cur = self._conn.execute(
            """SELECT s.id, s.name, s.kind, f.relative_path, s.start_line, s.parent_symbol
               FROM symbols s
               JOIN files f ON s.file_id = f.id
               WHERE s.name = ?
               ORDER BY s.start_line ASC""",
            (name,),
        )
        return [self._to_resolved(row) for row in cur.fetchall()]

    def _qualified_match(self, method: str, parent: str) -> list[ResolvedSymbol]:
        cur = self._conn.execute(
            """SELECT s.id, s.name, s.kind, f.relative_path, s.start_line, s.parent_symbol
               FROM symbols s
               JOIN files f ON s.file_id = f.id
               JOIN symbols p ON p.id = CAST(s.parent_symbol AS INTEGER)
               WHERE s.name = ? AND p.name = ?""",
            (method, parent),
        )
        return [self._to_resolved(row) for row in cur.fetchall()]

    def _file_match(self, name: str, file_part: str) -> list[ResolvedSymbol]:
        cur = self._conn.execute(
            """SELECT s.id, s.name, s.kind, f.relative_path, s.start_line, s.parent_symbol
               FROM symbols s
               JOIN files f ON s.file_id = f.id
               WHERE s.name = ? AND f.relative_path LIKE ?""",
            (name, f"%{file_part}%"),
        )
        return [self._to_resolved(row) for row in cur.fetchall()]

    @staticmethod
    def _to_resolved(row: tuple) -> ResolvedSymbol:
        parent_id = int(row[5]) if row[5] and str(row[5]).isdigit() else None
        return ResolvedSymbol(
            id=row[0], name=row[1], kind=row[2],
            file_path=row[3], line=row[4], parent_symbol_id=parent_id,
        )

File: backend/src/engine/graph/symbol-resolver.ts
/**
 * KSA-154: Symbol Resolver — resolves symbol names to database records.
 * Supports exact match, qualified names (Class.method), and file:symbol format.
 */

import Database from 'better-sqlite3';

export interface ResolvedSymbol {
  id: number;
  name: string;
  kind: string;
  filePath: string;
  line: number;
  parentSymbolId: number | null;
}

export class SymbolResolver {
  private db: Database.Database;
  private stmts: {
    exactMatch: Database.Statement;
    qualifiedMatch: Database.Statement;
    fileMatch: Database.Statement;
    fuzzyMatch: Database.Statement;
  };

  constructor(db: Database.Database) {
    this.db = db;
    this.stmts = {
      exactMatch: db.prepare(`
        SELECT s.id, s.name, s.kind, f.relative_path as filePath, s.start_line as line, s.parent_symbol_id as parentSymbolId
        FROM symbols s
        JOIN files f ON s.file_id = f.id
        WHERE s.name = ?
        ORDER BY s.start_line ASC
      `),
      qualifiedMatch: db.prepare(`
        SELECT s.id, s.name, s.kind, f.relative_path as filePath, s.start_line as line, s.parent_symbol_id as parentSymbolId
        FROM symbols s
        JOIN files f ON s.file_id = f.id
        JOIN symbols p ON p.id = s.parent_symbol_id
        WHERE s.name = ? AND p.name = ?
      `),
      fileMatch: db.prepare(`
        SELECT s.id, s.name, s.kind, f.relative_path as filePath, s.start_line as line, s.parent_symbol_id as parentSymbolId
        FROM symbols s
        JOIN files f ON s.file_id = f.id
        WHERE s.name = ? AND f.relative_path LIKE ?
      `),
      fuzzyMatch: db.prepare(`
        SELECT DISTINCT s.name
        FROM symbols s
        WHERE s.name LIKE ?
        LIMIT ?
      `),
    };
  }

  /** Resolve a symbol name to one or more database records. */
  resolve(input: string): ResolvedSymbol[] {
    // Strategy 1: Exact match
    let results = this.stmts.exactMatch.all(input) as ResolvedSymbol[];
    if (results.length > 0) return results;

    // Strategy 2: Qualified name (Class.method)
    if (input.includes('.')) {
      const dotIndex = input.lastIndexOf('.');
      const parent = input.substring(0, dotIndex);
      const method = input.substring(dotIndex + 1);
      results = this.stmts.qualifiedMatch.all(method, parent) as ResolvedSymbol[];
      if (results.length > 0) return results;
    }

    // Strategy 3: file:symbol format
    if (input.includes(':')) {
      const colonIndex = input.lastIndexOf(':');
      const file = input.substring(0, colonIndex);
      const name = input.substring(colonIndex + 1);
      results = this.stmts.fileMatch.all(name, `%${file}%`) as ResolvedSymbol[];
      if (results.length > 0) return results;
    }

    return [];
  }

  /** Suggest similar symbol names for "did you mean?" responses. */
  suggest(input: string, limit: number = 5): string[] {
    const rows = this.stmts.fuzzyMatch.all(`%${input}%`, limit) as { name: string }[];
    return rows.map(r => r.name);
  }
}
